src/lib/pubkey/ec_group/ec_apoint.cpp

```diff
@@ -111,12 +111,26 @@ EC_AffinePoint EC_AffinePoint::mul(const EC_Scalar& scalar, RandomNumberGenerato
    return EC_AffinePoint(inner().mul(scalar._inner(), rng, ws));
 }
 
-EC_AffinePoint EC_AffinePoint::mul_px_qy(const EC_AffinePoint& p,
-                                         const EC_Scalar& x,
-                                         const EC_AffinePoint& q,
-                                         const EC_Scalar& y,
-                                         RandomNumberGenerator& rng) {
+std::optional<EC_AffinePoint> EC_AffinePoint::mul_px_qy(const EC_AffinePoint& p,
+                                                        const EC_Scalar& x,
+                                                        const EC_AffinePoint& q,
+                                                        const EC_Scalar& y,
+                                                        RandomNumberGenerator& rng) {
    auto pt = p._inner().group()->mul_px_qy(p._inner(), x._inner(), q._inner(), y._inner(), rng);
+   if(pt) {
+      return EC_AffinePoint(std::move(pt));
+   } else {
+      return {};
+   }
```

Comment:
Perhaps its worth adding a utility for this pattern at one point. Something along those lines:

```C++
template <typename InnerT, typename T>
   requires std::constructible_from<InnerT, T> && requires(T t) {
      { static_cast<bool>(t) } -> std::same_as<bool>;
   }
auto maybe(T&& v) {
   std::optional<InnerT> result = std::nullopt;
   if(v) {
      result.emplace(std::forward<T>(v));
   }
   return result;
}
```

... the implementation of this function would then look like that:

```suggestion
   return maybe<EC_AffinePoint>(p._inner().group()->mul_px_qy(p._inner(), x._inner(), q._inner(), y._inner(), rng));
```

... and it would also be RVO-friendlier.

One downside: similarly to `std::make_unique` (and friends), this needs the constructor of `InnerT` (here: `EC_AffinePoint(std::unique_ptr<EC_AffinePoint_Data>)`) to be public. 😞 This would currently not be the case for this particular example.

Reply:
#4450